### DISCUSSION TITLE
Resolving preset for adding an option

### DISCUSSION THREAD

**Discussion by rea-ducks** [blakeblackshear/frigate#6639] [General]
Hi, I want to continue #6634 here, since it's more a general question too. My frigate config can be seen over there.

**Situation**
My camera being mounted rotated I've to correct the feed, sadly can't do this in camera. And I finally want to use Hardware acceleration (Skylake iGPU). In #6634 I learned, that I just can't use `hwaccel_args: preset-vaapi` and add `-vf transpose=cclock` in the output_args, since the [preset](https://github.com/blakeblackshear/frigate/blob/master/frigate/ffmpeg_presets.py) already contains -vf.

**What I've done**
My first try was to replace the preset by `hwaccel_args: -hwaccel vaapi -hwaccel_device /dev/dri/renderD128 -hwaccel_output_format yuv420p` in the global ffmpeg section. While this works and gives some hardware acceleration, the CPU usage is still fairly high (20% even without transpose vs. 2% with the full preset).
So there's still something missing, supposedly at least HW acceleration for encoding. I suppose it's necessary to resolve the presets and put the options at the right places into my config.
There's `PRESETS_HW_ACCEL_DECODE`, `PRESETS_HW_ACCEL_ENCODE` and `PRESETS_HW_ACCEL_SCALE`. But where in my config do I have to put those?

Any help is appreciated.

**Comment by NickM-27** [comment]
You would want your hwaccel_args to be `-hwaccel vaapi -hwaccel_device /dev/dri/renderD128 -hwaccel_output_format vaapi` and then your record output args -vf section to be `-vf vf format=vaapi|nv12,hwupload`

you would then need to play with the transpose to get the filter graph to work. That being said, vaapi does support vaapi_transpose which is done on the GPU so that may be worth looking into as well. go2rtc supports all of this without much effort so that would probably be the much easier way to go about it

**Reply by rea-ducks** [reply]
Using go2rtc was might first thought also. But then I thought, that would result in two recodings.
One from H.265 to H.264 (and transposing) by go2rtc and then in frigate using that stream and recoding it for Home Assistant Integration to display it via jsmpeg. Or am I wrong?

And additionally while following works:
```
go2rtc:
  streams:
    eyethree_restream:
      - ffmpeg:rtsp://192.168.8.62:554/channel=1&stream=1&protocol=unicast&onvif=0.sdp#video=h264#hardware=vaapi
```
as does 
`…
      - ffmpeg:rtsp://192.168.8.62:554/channel=1&stream=1&protocol=unicast&onvif=0.sdp#video=h264#rotate=270`
as well as
`…
      - ffmpeg:rtsp://192.168.8.62:554/channel=1&stream=1&protocol=unicast&onvif=0.sdp#video=h264#raw=-vf transpose=cclock`

But any rotation together with hardware results in an error, e.g.
`20:30:37.863 ERR github.com/AlexxIT/go2rtc/cmd/streams/producer.go:53 > error="exec: exit status 1" url=ffmpeg:rtsp://192.168.8.62:554/channel=1&stream=1&protocol=unicast&onvif=0.sdp#video=h264#hardware=vaapi#rotate=270
[2023-05-29 20:30:39] frigate.video                  ERROR   : eyethree: Unable to read frames from ffmpeg process.`

Judging from the CPU load, go2rtc uses vaapi also when not saying `#hardware`, but [go2rtc documentation on hardware acceleration](https://github.com/AlexxIT/go2rtc/wiki/Hardware-acceleration) says
> Acceleration is disabled by default

I'm not sure what to make of that.

**Reply by NickM-27** [reply]
No that is incorrect, frigate integration uses go2rtc if setup correctly. 

Also you shouldn't use raw args you should use `#rototate=90#hardware`

**Reply by rea-ducks** [reply]
As I wrote, any rotation together with hardware results in an error:
```
frigate9  | 2023-05-30 18:05:41.470615482  [INFO] Starting go2rtc healthcheck service...
frigate9  | 2023-05-30 18:05:43.869408585  18:05:43.869 ERR github.com/AlexxIT/go2rtc/cmd/streams/producer.go:53 > error="exec: exit status 1" url=ffmpeg:rtsp://192.168.8.62:554/channel=1&stream=1&protocol=unicast&onvif=0.sdp#video=h264#rotate=270#hardware
frigate9  | 2023-05-30 18:05:43.869412799  18:05:43.869 WRN [rtsp] error="source 0 error: exec: exit status 1" stream=eyethree_restream
frigate9  | 2023-05-30 18:05:48.245253384  [2023-05-30 18:05:48] frigate.record                 ERROR   : Error occurred when attempting to maintain recording cache
frigate9  | 2023-05-30 18:05:48.245662471  [2023-05-30 18:05:48] frigate.record                 ERROR   : list index out of range
frigate9  | 2023-05-30 18:05:53.103490055  [2023-05-30 18:05:53] watchdog.eyethree              ERROR   : Ffmpeg process crashed unexpectedly for eyethree.
frigate9  | 2023-05-30 18:05:53.103494333  [2023-05-30 18:05:53] watchdog.eyethree              ERROR   : The following ffmpeg logs include the last 100 lines prior to exit.
frigate9  | 2023-05-30 18:05:53.103496678  [2023-05-30 18:05:53] ffmpeg.eyethree.detect         ERROR   : rtsp://127.0.0.1:8554/eyethree_restream: Invalid data found when processing input
frigate9  | 2023-05-30 18:05:53.159292051  [2023-05-30 18:05:53] frigate.record                 ERROR   : Error occurred when attempting to maintain recording cache
frigate9  | 2023-05-30 18:05:53.159741251  [2023-05-30 18:05:53] frigate.record                 ERROR   : list index out of range
frigate9  | 2023-05-30 18:05:58.223391154  [2023-05-30 18:05:58] frigate.video                  ERROR   : eyethree: Unable to read frames from ffmpeg process.
frigate9  | 2023-05-30 18:05:58.223395134  [2023-05-30 18:05:58] frigate.video                  ERROR   : eyethree: ffmpeg process is not running. exiting capture thread...
frigate9  | 2023-05-30 18:05:58.223745462  [2023-05-30 18:05:58] frigate.record                 ERROR   : Error occurred when attempting to maintain recording cache
frigate9  | 2023-05-30 18:05:58.223748308  [2023-05-30 18:05:58] frigate.record                 ERROR   : list index out of range
frigate9  | 2023-05-30 18:05:58.403966454  18:05:58.403 ERR github.com/AlexxIT/go2rtc/cmd/streams/producer.go:53 > error="exec: exit status 1" url=ffmpeg:rtsp://192.168.8.62:554/channel=1&stream=1&protocol=unicast&onvif=0.sdp#video=h264#rotate=270#hardware
```

**Reply by NickM-27** [reply]
You should enable trace logs for exec so we can see what the error is

**Reply by trinitronx** [reply]
Just found this in search results while trying to get `transpose_vaapi` working, and figured I should share what worked for me with plain `ffmpeg` commands:

```shell
# Command to correct a 180° flipped iPhone video (`.mov`, HEVC 10-bit)
ffmpeg -threads $(nproc --all) -init_hw_device vaapi=intel0:/dev/dri/renderD128 \
  -hwaccel vaapi -hwaccel_output_format vaapi -hwaccel_device intel0 -filter_hw_device intel0 \
  -i input.mov  \
    -vf 'format=vaapi,hwupload,transpose_vaapi=clock,transpose_vaapi=clock' \
    -map 0:0 -c:v hevc_vaapi -global_quality 25 -profile:v:0 main10 \
    -map 0:a -c:a copy -strict strict \
  -f mp4 -y  out.rotated.mov
```

The important bits are:

- The VAAPI hardware init flags: `-init_hw_device vaapi=intel0:/dev/dri/renderD128 -hwaccel vaapi -hwaccel_output_format vaapi -hwaccel_device intel0 -filter_hw_device intel0`
- Use `-vf format=vaapi,hwupload, ... ` **first** before your filters.  (`format` must be one supported by the hardware, e.g.: `nv12`)
- Filters with `*_vaapi` suffix run on the GPU, while filters without this run on the CPU.  (Same idea for `*_cuda`, `*_opencl`, `*_qsv`, etc...)

For debugging `ffmpeg`, the `-report` flag is helpful.  It will write a report like `ffmpeg-${date}-${pid}.log` to the current working directory of the command.  In this, you'll see how it's parsing the options, whether it initializes VAAPI drivers properly, and whether driver and hardware supports your codecs, filters, etc...

Also: It didn't work until my system had the correct drivers (there are 2 which both supported my Kaby-Lake hardware: `intel-media-driver`, a.k.a. `iHD_drv_video.so` **AND:** `libva-intel-driver`, a.k.a. `i965_drv_video.so`).  Also, for h.265 / HEVC support, I needed to [enable Intel GuC/HuC firmware][1] via:

```
# /etc/modprobe.d/intel-i915-hd-graphics-930.conf
# 2023-05-05: Configuring some intel i915 kernel module options
#             to test integrated Intel HD Graphics 630 + i7-7700K Kaby Lake CPU + Z170 Chipset
# Enable Intel GuC and HuC and Framebuffer compression
# Source: https://discussion.fedoraproject.org/t/intel-graphics-best-practices-and-settings-for-hardware-acceleration/69944
# Reference: https://gist.github.com/Brainiarc7/aa43570f512906e882ad6cdd835efe57
options i915 enable_guc=2
# https://wiki.archlinux.org/title/intel_graphics#Framebuffer_compression_(enable_fbc)
options i915 enable_fbc=1
```

Some important things for using `*_vaapi` filters on Intel hardware:

  -  [`intel-media-driver`][2] / `iHD_drv_video.so` is the only driver which supported applying shaders / filters such as `transpose_vaapi`, `denoise_vaapi`, etc...
     - Intel recommends this for media processing
  - [`libva-intel-driver`][3] / `i965_drv_video.so` supports decode & encode, and **only** simple filters like `scale_vaapi`  (at least on the hardware I tested)
    - If you have only this installed, `ffmpeg` will say your driver does not support the more complex `*_vaapi` filters.

[1]: https://wiki.archlinux.org/title/intel_graphics#Enable_GuC_/_HuC_firmware_loading
[2]: https://github.com/intel/media-driver
[3]: https://github.com/intel/intel-vaapi-driver
